src/fastpath/fastpath.c

```diff
@@ -15,6 +15,289 @@
+                 !cap_endpoint_cap_get_capCanReceive(handler_cap))) {
+        vm_fault_slowpath(type);
+    }
+
+    /* Get the endpoint address */
+    ep_ptr = EP_PTR(cap_endpoint_cap_get_capEPPtr(handler_cap));
+
+    /* Get the destination thread, which is only going to be valid
+    * if the endpoint is valid. */
+    dest = TCB_PTR(endpoint_ptr_get_epQueue_head(ep_ptr));
+
+    /* Check that there's a thread waiting to receive */
+    if (unlikely(endpoint_ptr_get_state(ep_ptr) != EPState_Recv)) {
+        vm_fault_slowpath(type);
+    }
+
+    /* ensure we are not single stepping the destination in ia32 */
+#if defined(CONFIG_HARDWARE_DEBUG_API) && defined(CONFIG_ARCH_IA32)
+    if (unlikely(dest->tcbArch.tcbContext.breakpointState.single_step_enabled)) {
+        vm_fault_slowpath(type);
+    }
+#endif
+
+    /* Get destination thread.*/
+    newVTable = TCB_PTR_CTE_PTR(dest, tcbVTable)->cap;
+
+    /* Get vspace root. */
+    cap_pd = cap_vtable_cap_get_vspace_root_fp(newVTable);
+
+    /* Ensure that the destination has a valid VTable. */
+    if (unlikely(! isValidVTableRoot_fp(newVTable))) {
+        vm_fault_slowpath(type);
+    }
+
+
+#ifdef CONFIG_ARCH_AARCH32
+    /* Get HW ASID */
+    stored_hw_asid = cap_pd[PD_ASID_SLOT];
+#endif
+
+#ifdef CONFIG_ARCH_X86_64
+    /* borrow the stored_hw_asid for PCID */
+    stored_hw_asid.words[0] = cap_pml4_cap_get_capPML4MappedASID_fp(newVTable);
+#endif
+
+#ifdef CONFIG_ARCH_IA32
+    /* stored_hw_asid is unused on ia32 fastpath, but gets passed into a function below. */
+    stored_hw_asid.words[0] = 0;
+#endif
+
+#ifdef CONFIG_ARCH_AARCH64
+    stored_hw_asid.words[0] = cap_vtable_root_get_mappedASID(newVTable);
+#endif
+
+#ifdef CONFIG_ARCH_RISCV
+    /* Get HW ASID */
+    stored_hw_asid.words[0] = cap_page_table_cap_get_capPTMappedASID(newVTable);
+#endif
+
+    /* let gcc optimise this out for 1 domain */
+    dom = maxDom ? ksCurDomain : 0;
+    /* ensure only the idle thread or lower prio threads are present in the scheduler */
+    if (unlikely(dest->tcbPriority < NODE_STATE(ksCurThread->tcbPriority) &&
+                 !isHighestPrio(dom, dest->tcbPriority))) {
+
+        vm_fault_slowpath(type);
+    }
+
+    /* Ensure that the endpoint has has grant or grant-reply rights so that we can
+    * create the reply cap */
+    if (unlikely(!cap_endpoint_cap_get_capCanGrant(handler_cap) &&
+                 !cap_endpoint_cap_get_capCanGrantReply(handler_cap))) {
+        vm_fault_slowpath(type);
+    }
+
+#ifdef CONFIG_ARCH_AARCH32
+    if (unlikely(!pde_pde_invalid_get_stored_asid_valid(stored_hw_asid))) {
+        vm_fault_slowpath(type);
+    }
+#endif
+
+    /* Ensure the original caller is in the current domain and can be scheduled directly. */
+    if (unlikely(dest->tcbDomain != ksCurDomain && 0 < maxDom)) {
+        vm_fault_slowpath(type);
+    }
+
+#ifdef CONFIG_KERNEL_MCS
+    if (unlikely(dest->tcbSchedContext != NULL)) {
```

Comment:
So we're only fastpathing the case where the scheduling context is donated? does that make sense in the context of a VM fault fastpath?

Reply:
I think when I discussed this with Gernot and Kevin, we agreed that passive servers make sense for a lot of fault handlers. Additionally, I based a lot of this logic off the seL4_Call fastpath, which also only supports calling to passive servers at the moment. I can see that it would be ideal to be able to support active servers as fault handlers too, but I would need to have a better understanding of why this is not an option for seL4_Call and what could be done to make it work for this fastpath.

Reply:
it doesn't make sense for seL4_Call as you need to pull in nearly the whole scheduler if you are switching SC's, as at minimum you need to reprogram the timer and check if you need to wake other tasks. There are subsets of that that could be pulled in, but you'd need to choose which scheduling scenario you want to fastpath. 

Also, the less we pull in the easier it is for verification to verify.